internal/topology/variables/patch_validation.go

```diff
@@ -0,0 +1,222 @@
+the template and validating if the result is a valid YAML or JSON value.
+
+*/
+func validatePatchDefinitions(patch clusterv1.ClusterClassPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, definition := range patch.Definitions {
+		allErrs = append(allErrs,
+			validatePatchDefinitionSelector(definition.Selector, path)...)
+
+		allErrs = append(allErrs,
+			validatePatchJSONPatches(definition.JSONPatches, variables, path)...)
+	}
+	return allErrs
+}
+
+func validatePatchJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	validOps := sets.NewString("add", "replace", "remove")
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("op").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch op must be one of \"add\" \"remove\" or \"replace\". %v is not valid.", jsonPatch.Op),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		if _, err := jsonptr.Parse(jsonPatch.Path); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("path").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch path must be a valid JSON pointer (RFC 6901). %s not valid.", jsonPatch.Path),
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of Value or ValueFrom.",
+				))
+		}
+		if jsonPatch.Value != nil && jsonPatch.ValueFrom != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch can not define both Value and ValueFrom.",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template == nil && jsonPatch.ValueFrom.Variable == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"ValueFrom must set either \"/Template or Variable/\".",
+				))
+		}
+		if jsonPatch.ValueFrom != nil && jsonPatch.ValueFrom.Template != nil && jsonPatch.ValueFrom.Variable != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"ValueFrom can not set both \"/Template and Variable/\".",
+				))
+		}
+
+		//TODO: return early if the jsonPatch has the "replace" op. Additional validation is not needed.
+		// if jsonPatch.Op == "replace" {
+		//	return allErrs
+		//}
```

Comment:
Uncomment (or remove)

Reply:
I think it should be remove. 

I think we need the following validation:
* add: index `0` and `-` is allowed
* remove/replace: no index is allowed

+ I wonder if we should move the variable validation in a separate func and only call it if op is add or replace.

Reply:
Here the code returns early i it's remove, and (if we want to keep the structure) we'd return early again here if the op is not add. Alternatively we can break some of this out into seperate functions.

Reply:
I meant remove this code with "it should be remove" :)

But independent of that. I think we need validation of the path in all 3 cases and it's different. If there's a good way to do it inline that's fine for me, separate funcs as well.

Reply:
The new version has a function which handles the two cases.